### ISSUE TITLE
Network model

### ISSUE THREAD

**Issue by ralphevins** [hues-platform/python-ehub#90]
I'd like to try a simple test of a network of hubs.

General outline:
1. Add a placeholder energy stream called 'network' that appears to provide free electricity (coefficient of 1 in the elec stream, no -1 coefficient)
2. Instantiate two different hub models
3. Merge them into one common LP
4. Add a custom constraint that sets Hub1network = - Hub2network
5. Test it on a case where Hub1 has lots of PV at some times and Hub2 at other times to see if they share.

The issue with (3) will be to avoid LP variable naming conflicts.
Could prepend the hub name or ID to everything to avoid this.

**Comment by ralphevins** [comment]
@greyhillman

**Comment by greyhillman** [comment]
I don't think there will be any variable naming conflicts since the name they get is from a class variable that increments on everything call to get a new name.
I'll take a shot at this tomorrow.

**Comment by greyhillman** [comment]
This will have to wait until #87 is merged. Pyomo doesn't have a nice way of "merging" two linear programming problems. With PyLP (the new wrapper) would make this pretty easy to do. You would take the two list of constraints from each instance and add them with the custom constraint. The new objective function would then just be the addition of both instance's objective function.

**Comment by greyhillman** [comment]
I was able to merge the two LP problems. The problems were independent though.

We would have to meet in order to discuss how to get one model to send stuff to the other and vice versa. There's some limitations on the current model for doing such a thing.

**Comment by ralphevins** [comment]
Great that the merge worked.
Is there an issue with my suggested approach for coupling the models (steps 1 and 4 in this issue)?
Would also need to merge the objective functions, can use a simple sum for now.

**Comment by greyhillman** [comment]
I'm more confused on what to do with Step 4.

I tried adding another stream called `Network` with a converter that uses that stream as an input (since the converter has to have at least one input (due to how the code works)). Then I tried to figure out a constraint that links the two models. I think that's where the problem lies.

**Comment by ralphevins** [comment]
Need to make sure we're not enforcing an energy balance internally for the network stream or it will force it to zero, since when we build that hub there's only one entry.
It might also be that the stream has to be allowed to go negative, unlike the others?

So you have:
Hub1network -> Elec in hub 1  (using the dummy converter you specified)
Hub2network -> Elec in hub 2
Hub1network = - Hub2network (custom constraint added later)

That way if there's say 10kW of PV available in hub 1 but no demand, and 10kW demand in hub 2 but no PV, you should get:
Hub1PV = 10 
Hub1network = -10 (consuming electricity from PV, so the elec stream sums to zero)
Hub2network = 10 (via the constraint)
Hub2demand = -10 (consuming electricity from network, so elec stream sums to zero)

**Comment by greyhillman** [comment]
What's confusing me is the `Hub1network = - Hub2network` part. Is this like the energy imported for the network converter per time step?

**Comment by ralphevins** [comment]
Yep. So Hub1network is energy received by hub1 from the network (+ve) or exported to the network (-ve). Hence the need to refer to the two 'network' variables of the two LPs by some sort of ID (you said it's a class variable that gets incremented? Might need you to explain that.)
Looking forwards, a more generic way to do it would be to say that the sum of all 'network' variables has to sum to zero, giving network continuity for lots of hubs.
But another issue is that we'd like to be able to specify connectivity (hub 1-> hub2 but not hub 1 -> hub 3) and also network losses e.g. (hub1network = -0.95 * hub2network)...

**Comment by greyhillman** [comment]
It works!

I don't have time to work on a networked version of the model though. My only suggestion is
1) Work out the model in Math first. Like create a network version of the energy hub model and all the equations for it.
2) After the Math is good, then try and build the Python around it. Have it easy to connect models.
The biggest thing I learned from the Beyond PEP 8 video was to write the code how you want to write it first. Then write the code to implement it in the same way. For example, having a list of tuples of the hub objects and then have a function that does the work of connecting two of them together.
